### ISSUE TITLE
Long Polyline rendering bug: Shape path too large to be rendered into a texture

### ISSUE THREAD

**Issue by stellardeveloper** [MKergall/osmbonuspack#171]
There is one more  problem with long Polylines:
Depending on length and orientation, the line will disappear from a certain zoom level upward. 
In the logcat you will find something like;: 
"W/OpenGLRenderer(29239): Shape path too large to be rendered into a texture". 
The reason is, that for rendering the path there has to be created a kind of bitmap in memory containing all the pixels. For long polylines at higher zoom levels this bitmap will incease unreasonably large. I tried an extreme example: non-geodetic line from (90, 90) to (-90, -90), which will disappear at zoomLevel 3, I found two solution recipes in the internet: 
1) < application android:hardwareAccelerated="false" > This did not fix the problem
2) Drawing the path to a temporary bitmap in screen size, then draw the bitmap to canvas.

I tried the follwing code in the Polyline.draw:

```
    final Rect dest = new Rect();
    dest.set(0, 0, mapView.getWidth()-1,  mapView.getHeight()-1);
    final Bitmap.Config conf = Bitmap.Config.ARGB_8888; // see other conf types
    final Bitmap bmp = Bitmap.createBitmap(mapView.getWidth(), mapView.getHeight(), conf); // this creates a MUTABLE bitmap
    final Canvas bmpcanvas = new Canvas(bmp);
    if (mapView.getMapOrientation() != 0.0f) bmpcanvas.rotate(mapView.getMapOrientation(),  mapView.getWidth()/2,  mapView.getHeight()/2);
    bmpcanvas.drawPath(mPath, mPaint);

         // .. repeatPath code segment  following ...

    if (mapView.getMapOrientation() != 0.0f) canvas.rotate(-mapView.getMapOrientation(), mapView.getWidth()/2,  mapView.getHeight()/2);
    canvas.drawBitmap (bmp, dest, dest, null);
    if (mapView.getMapOrientation() != 0.0f) canvas.rotate(mapView.getMapOrientation(), mapView.getWidth()/2,  mapView.getHeight()/2);
```

This is working,  but slowing down the rendering extremely. I think we have to insert a check for points exceeding the screen boundary, which should not be sent to the canvas.

**Comment by stellardeveloper** [comment]
My fix for this issue 171, and issues [170] (https://github.com/MKergall/osmbonuspack/issues/170) and [169] (https://github.com/MKergall/osmbonuspack/issues/170):

	protected void draw (final Canvas canvas, final MapView mapView, final boolean shadow) {

		if (shadow) {
			return;
		}

		final int size = mPoints.size();
		if (size < 2) {
			// nothing to paint
			return;
		}
		
		final Projection pj = mapView.getProjection();
		// 180° in longitude in pixels, for avoiding lines > 180° in length
		final int halfMapSize = TileSystem.MapSize(mapView.getProjection().getZoomLevel())/2; 
		 // southern Limit of the map in Pixels, for detecting map crossing lines
		final int southLimit  = pj.toPixelsFromMercator(0, halfMapSize*2, null).y-1;           
		// calculate clipBounds in screen coordinates for removing "invisible" line segments
		final Rect clipBounds = new Rect(0, 0, mapView.getWidth ()-1, mapView.getHeight()-1);
		// take into account map orientation:
		if (mapView.getMapOrientation() != 0.0f)
			GeometryMath.getBoundingBoxForRotatatedRectangle(clipBounds, mapView.getMapOrientation(), clipBounds);
		
		// area of visible line segments
		Rect pathExtension = new Rect(clipBounds); // used for decision: canvas.draw (fast) or bitmap draw (slow, avoiding OpenGLRenderer problem) 
		boolean lineVisible = false;			

		// precompute new points to the intermediate projection.
		precomputePoints(pj);

		final Point projectedPoint0 = mPoints.get(0); // points from the points list

		final Point screenPoint0 = pj.toPixelsFromProjected(projectedPoint0, mTempPoint1); // points on screen
		Point projectedPoint1;
		Point screenPoint1 ;

		mPath.rewind();

		for (int i = 1; i < size; i++) {			

			// compute next points
			projectedPoint1 = mPoints.get(i);
			screenPoint1 = pj.toPixelsFromProjected(projectedPoint1, this.mTempPoint2);
			if (Math.abs(screenPoint1.x - screenPoint0.x) + Math.abs(screenPoint1.y - screenPoint0.y) <= 1
					&& screenPoint0.y != southLimit && screenPoint1.y != southLimit) { // do not skip points  crossing maps! 
				// skip this point, too close to previous point
				continue;
			}
			
			
			// check for lines exceeding 180° in longitude, cut line into two segments
			if  (  (Math.abs(screenPoint1.x - screenPoint0.x) > halfMapSize) 
			// check for lines crossing the southern limit
					|| (screenPoint1.y > southLimit) != (screenPoint0.y > southLimit)) {
				// handle x and y coordinates separately
				int x0 = screenPoint0.x;
				int y0 = screenPoint0.y;
				int x1 = screenPoint1.x;
				int y1 = screenPoint1.y;

				// first check x
				if  (Math.abs(screenPoint1.x - screenPoint0.x) > halfMapSize) {// x has to be adjusted 	
					if (screenPoint1.x < mapView.getWidth() /2) { 
						// screenPoint1 is left of screenPoint0
						x1 += halfMapSize*2; // set x1 360° east of screenPoint1
						x0 -= halfMapSize*2; // set x0 360° west of screenPoint0
					} else {
						x1 -= halfMapSize*2;
						x0 += halfMapSize*2;
					}								
				}
							
				// now check y
				if ((screenPoint1.y > southLimit) != (screenPoint0.y > southLimit)) {
					// line is crossing from one map to the other
					if (screenPoint1.y > southLimit) {						
						// screenPoint1 was switched to map below
						y1 -= halfMapSize*2;  // set y1 into northern map
						y0 += halfMapSize*2;  // set y0 into map below 						
					} else {
						y1 += halfMapSize*2;
						y0 -= halfMapSize*2;
					}				
				}
				// create rectangle of line segment, ensure top < bottom and right < left  to obtain valid rectangle!
				mLineBounds.set(Math.min(screenPoint0.x, x1), Math.min(screenPoint0.y, y1),
						Math.max(screenPoint0.x, x1), Math.max(screenPoint0.y, y1)); 
				// check whether this line segment is visible
				if (Rect.intersects(clipBounds, mLineBounds)) {
					mPath.moveTo(screenPoint0.x, screenPoint0.y);
					mPath.lineTo(x1, y1);  
					pathExtension.union(mLineBounds); // caution! buggy for invalid rectangles (top > bottom or right > left) !
					lineVisible = true;
				}
				screenPoint0.x = x0;
				screenPoint0.y = y0;
			} // end of line break check
			
			// check, whether this line segment is visible, ensure top < bottom and right < left  to obtain valid rectangle!
			mLineBounds.set(Math.min(screenPoint0.x, screenPoint1.x), Math.min(screenPoint0.y, screenPoint1.y),
					Math.max(screenPoint0.x, screenPoint1.x), Math.max(screenPoint0.y, screenPoint1.y));					
			if (Rect.intersects(clipBounds, mLineBounds)) {
				mPath.moveTo(screenPoint0.x, screenPoint0.y);
				mPath.lineTo(screenPoint1.x, screenPoint1.y);
				pathExtension.union(mLineBounds);
				lineVisible = true;
			} 

			// update starting point to next position
			screenPoint0.x = screenPoint1.x;
			screenPoint0.y = screenPoint1.y;
		}
		
		
		// send only visible lines to canvas, not segments outside screen area
		if (lineVisible) {

			//  check, whether visible line segments cover less than 2048 pixels rectangle. 
			if (Math.max(pathExtension.width(), pathExtension.height()) <= 2048) {
				//  Use fast canvas.drawPath method.
				canvas.drawPath(mPath, mPaint);
			} else {

				// fixing W/OpenGLRenderer(29239): Shape path too large to be rendered into a texture.
				// This will occur, if memory exceeds 2040 pixels => Path will disappear.
				// Draw path into temporary bitmap in screen size, then send bitmap to screen canvas.	
				// <application android:hardwareAccelerated="false" > does not fix the OpenGLRenderer problem!
				// Line (80, 80) to (-80, 80) will disappear at zoomLevel 4, Line (90, 90) to (-90, 90) at zoomLevel 3
				final Rect dest = new Rect(0, 0, mapView.getWidth()-1,  mapView.getHeight()-1);
				//			dest.set(0, 0, mapView.getWidth()-1,  mapView.getHeight()-1);
				//			final Bitmap.Config conf = Bitmap.Config.ARGB_8888; // 8 bit Aplpha, RGB
				final Bitmap bmp = Bitmap.createBitmap(mapView.getWidth(), mapView.getHeight(), Bitmap.Config.ARGB_8888); // this creates a MUTABLE bitmap
				final Canvas bmpcanvas = new Canvas(bmp);
				// Draw path to bitmap according to actual screen rotation
				if (mapView.getMapOrientation() != 0.0f) bmpcanvas.rotate(mapView.getMapOrientation(),  mapView.getWidth()/2,  mapView.getHeight()/2);
				bmpcanvas.drawPath(mPath, mPaint);

				// Draw bitmap according to actual screen rotation, derotate canvas if necessary
				if (mapView.getMapOrientation() != 0.0f) canvas.rotate(-mapView.getMapOrientation(), mapView.getWidth()/2,  mapView.getHeight()/2);
				canvas.drawBitmap (bmp, dest, dest, null);
				//restore canvas rotation if necessary
				if (mapView.getMapOrientation() != 0.0f) canvas.rotate(mapView.getMapOrientation(), mapView.getWidth()/2,  mapView.getHeight()/2);
			}
		}

**Comment by stellardeveloper** [comment]
I modified the above code for better  performance. Replace the time consuming bitmap creation

      final Bitmap mBmp= Bitmap.createBitmap(mapView.getWidth(), mapView.getHeight(), Bitmap.Config.ARGB_8888);

by recyling an already created bitmap:

				if (mBmp ==  null || mBmp.getWidth() != mapView.getWidth() || mBmp.getHeight() != mapView.getHeight()) { // create Bitmap only once if needed
					mBmp = Bitmap.createBitmap(mapView.getWidth(), mapView.getHeight(), Bitmap.Config.ARGB_8888); // this most time consuming step, app. 30 ms!
				} else mBmp.eraseColor(Color.TRANSPARENT); // recycle bitmap and erase the old bitmap, much faster 0.6 .. 0.9 ms

and declare  a public static Bitmap mBmp = null;variable in the Polyline class.

**Comment by MKergall** [comment]
Huge work, but more and more complex... I'm a little bit worried about the potential hidden issues. 
What about providing - at least temporarily - 2 draw methods (whith a flag to choose): the simple one and the advanced one? 

I was also rethinking about this "vertical replication" of the map (and of its overlays) : in fact, it's a non-sense. The north pole is not "just below" the south pole. 
Google Maps, OSM web site, Bing Maps, MapQuest, Mappy: nobody is doing that. They just put a grey filling. 
I will suggest to osmdroid team to remove this vertical replication of the map. In the meantime, we should not put code to handle it in overlays.

**Comment by stellardeveloper** [comment]
> What about providing - at least temporarily - 2 draw methods (whith a flag to choose): the simple one and the advanced one? 

I already thought of introducing it :-)
When testing the new draw method in the last days I got a good performance if I am using the static Bitmap mBmp. I did not find any other Polyline issues until now.
Therefore I think the default option should be using the complex method, with the option to fall back to the simple one if setting a (static?) variable like "useSimpleDraw" to true.

Perhaps the Polygon class shares the same problems? I did not use them until now, but this should be checked, too.

**Comment by MKergall** [comment]
Quick impact evaluation: 

All overlays share the issue about horizontal replication at low zoom level: Polygon, Marker, and all osmdroid native overlays. 
Some online maps (Bing, MapQuest) completely avoid this issue by setting a min zoom level to a value with no replication at all... 
Leaflet do not handle it. 
Google Maps is the only place where I've found full support for horizontal replication. 

Polygon certainly shares all Polyline issues: long segments, segments crossing 180° line. With even more complexity around holes. 

Polyline event detection method is strongly concerned by 180° line cut. 

I bet the Marker Clusterer has very strange behaviours if markers are around the 180° line. 

osmdroid BoundingBoxE6 will also have issues with 180° line.

**Comment by stellardeveloper** [comment]
Hi,
I noticed, that sometimes the pj.toPixelsFromProjected method returns y values projecting onto an invisible replicated map north of the actual map, causing similar problems like we discussed in issue [170] (https://github.com/MKergall/osmbonuspack/issues/170).
My solution: checking, whether points come north of the northern map border:

     final int northLimit  = southLimit-halfMapSize*2+1;       

and cecking screenpoint0 

    if (screenPoint0.y < northLimit) screenPoint0.y += 2*halfMapSize; 

respectively screenPoint1

      if (screenPoint1.y < northLimit) screenPoint1.y += 2*halfMapSize;

**Comment by spyhunter99** [comment]
I've been watching this thread closely as I've ran into issues with drawing lat/lon grid lines. There's been some recent changes to the projection.adjustForDateline function. 

My 2 cents on the "Shape path too large to be rendered into a texture" issue. When the lat/lon gridlines with FolderOverlay, the trick was to only tell osmdroid about multipoint things (lines, polygons, etc) when it's within the bounds of the screen.  To accomplish this, I added a map listener which triggers a method on zoom and movement to remove, recalculate and readd the overlay containing the gridlines. If there are things which span long distances, you have clip the item at or near the bounds of the current view, otherwise it seems like it tries to allocate a bitmap for the whole thing. I say "seem" because I'm not entirely sure where in the code these parts are handled nor I have i looked into it. An example: you have a line near the north pole and the map view is in africa at zoom=13. Remove the overlays from osmdroid entirely for the north pole lines.

Personally, I think osmdroid and any overlay that provides it's own draw method should handle this gracefully.

Vertical replication should be nuked, but I can't figure out where this is happening. I think it's in the projection class but i'm not entirely sure. 

solutions to any of these problems would be awesome
